src/ssl.c

```diff
@@ -46909,35 +47063,73 @@ void wolfSSL_CRYPTO_cleanup_all_ex_data(void){
 }
 
 
-#ifndef NO_WOLFSSL_STUB
 WOLFSSL_DH *wolfSSL_DH_generate_parameters(int prime_len, int generator,
                            void (*callback) (int, int, void *), void *cb_arg)
 {
-    (void)prime_len;
-    (void)generator;
+    WOLFSSL_DH* dh;
+
+    WOLFSSL_ENTER("wolfSSL_DH_generate_parameters");
     (void)callback;
     (void)cb_arg;
-    WOLFSSL_ENTER("wolfSSL_DH_generate_parameters");
-    WOLFSSL_STUB("DH_generate_parameters");
 
-    return NULL;
+    if ((dh = wolfSSL_DH_new()) == NULL) {
+        WOLFSSL_MSG("wolfSSL_DH_new error");
+        return NULL;
+    }
+
+    if (wolfSSL_DH_generate_parameters_ex(dh, prime_len, generator, NULL)
+            != WOLFSSL_SUCCESS) {
+        WOLFSSL_MSG("wolfSSL_DH_generate_parameters_ex error");
+        wolfSSL_DH_free(dh);
+        return NULL;
+    }
+
+    return dh;
 }
-#endif
 
-#ifndef NO_WOLFSSL_STUB
 int wolfSSL_DH_generate_parameters_ex(WOLFSSL_DH* dh, int prime_len, int generator,
                            void (*callback) (int, int, void *))
 {
-    (void)prime_len;
-    (void)generator;
-    (void)callback;
-    (void)dh;
+    DhKey* key;
+
     WOLFSSL_ENTER("wolfSSL_DH_generate_parameters_ex");
-    WOLFSSL_STUB("DH_generate_parameters_ex");
+    (void)callback;
 
-    return -1;
+    if (dh == NULL) {
+        WOLFSSL_MSG("Bad parameter");
+        return WOLFSSL_FAILURE;
+    }
+
+    if (initGlobalRNG == 0 && wolfSSL_RAND_Init() != WOLFSSL_SUCCESS) {
+        WOLFSSL_MSG("No RNG to use");
+        return WOLFSSL_FAILURE;
+    }
+
+    if (dh->inSet == 0) {
+        if (SetDhInternal(dh) != WOLFSSL_SUCCESS) {
+            WOLFSSL_MSG("Unable to set internal DH structure");
+            return WOLFSSL_FAILURE;
+        }
+    }
+
+    key = (DhKey*)dh->internal;
+    if (mp_set_int(&key->g, generator) != MP_OKAY) {
```

Comment:
This doesn't do anything.
mp_init_multi is called with dh->g and this leaks with integer.c.

Reply:
Right. I tried writing a function to set the generator and check its validity using Appendix A.2.2 of FIPS 186-4 but from testing it looks like only numbers generated from A.2.1 satisfy that condition. I opted to just ignore the input generator parameter and keep the one generated by wolfSSL.